packages/atclient/src/notification.c

```diff
@@ -0,0 +1,322 @@
+#include "atclient/notification.h"
+#include "atclient/connection.h"
+#include "atclient/constants.h"
+#include "atclient/stringutils.h"
+#include "cJSON/cJSON.h"
+#include <atchops/uuid.h>
+#include <atlogger/atlogger.h>
+#include <string.h>
+#include <sys/time.h>
+
+void atclient_notify_params_init(atclient_notify_params *params) {
+  params->message_type = NMT_key;
+  params->priority = NP_low;
+  params->strategy = NS_all;
+  params->latest_n = 1;
+  params->notifier = ATCLIENT_DEFAULT_NOTIFIER;
+  params->notification_expiry = 24 * 60 * 60 * 1000; // 24 hours in milliseconds
+}
+
+void atclient_notify_params_free(atclient_notify_params *params) {
+  atclient_atkey_free(&params->key);
+  free(params->value);
+  free(params->notifier);
+}
+
+int atclient_notify(atclient *ctx, atclient_notify_params *params) {
+  int res = 1;
+  // Step 1 calculate the buffer size needed for the protocol command
+  size_t bufsize = 6 + 1 + 2 + 1 + 36 + 3 +              // "notify:id:" + $uuid + trailing '\r\n\0'
+                   1 + 8 + 1 + strlen(params->notifier); // ":notifier:$notifier"
+
+  // minimum viable: notify:update:key:value -> id
+  // TODO: check for bad input in params struct since we don't provide the best helper functions
+  // and add valid ones to bufsize
+
+  if (params->operation != NO_none) {
+    bufsize += 1 + 6; // ":update" | ":delete"
+  }
+
+  if (params->message_type != NMT_none) {
+    bufsize += 1 + 11 + 1 + 4; // ":messageType" + ":text" | ":key"
+  }
+
+  if (params->priority != NP_none) {
+    bufsize += 1 + 8 + 1 + 6; // ":priority" + ":low" | ":medium" | ":high"
+  }
+
+  if (params->strategy != NS_none) {
+    bufsize += 1 + 8 + 1 + 6; // ":strategy" + ":all" | ":latest"
+  }
```

Comment:
For these enums, we assume the largest string length of the options, this at most costs an extra 4 bytes of memory, which I don't think is worth the additional computational overhead, especially since this is temporary call stack memory.